python/modules/GenPartFilter.py

```diff
@@ -0,0 +1,123 @@
+            self.statusFlags = np.array([0]*length)
+
+        if statusLen == length:
+            self.statuses = np.array(statuses)
+        else:
+            if statuses: print "status length does not match, ignoring status"
+            self.statuses = np.array([0]*length)
+
+        if pdgIdLen == length:
+            self.parentPdgIds = np.array(pdgIds)
+        else:
+            if pdgIds: print "pdgId length does not match, ignoring pdgId"
+            self.parentPdgIds = np.array([0]*length)
+
+        self.vector_recursiveMotherSearch = np.vectorize(self.recursiveMotherSearch, excluded=(1,2))
+
+    def beginJob(self):
+        pass
+    def endJob(self):
+        pass
+
+    def beginFile(self, inputFile, outputFile, inputTree, wrappedOutputTree):
+        self.out = wrappedOutputTree
+
+        self.out.branch("GenPartCut_eta",              "F", lenVar = "nGenPartCut", title=inputTree.GetBranch("GenPart_eta").GetTitle() )
+        self.out.branch("GenPartCut_mass",             "F", lenVar = "nGenPartCut", title=inputTree.GetBranch("GenPart_mass").GetTitle())
+        self.out.branch("GenPartCut_phi",              "F", lenVar = "nGenPartCut", title=inputTree.GetBranch("GenPart_phi").GetTitle())
+        self.out.branch("GenPartCut_pt",               "F", lenVar = "nGenPartCut", title=inputTree.GetBranch("GenPart_pt").GetTitle())
+        self.out.branch("GenPartCut_genPartIdxMother", "I", lenVar = "nGenPartCut", title=inputTree.GetBranch("GenPart_genPartIdxMother").GetTitle())
+        self.out.branch("GenPartCut_pdgId",            "I", lenVar = "nGenPartCut", title=inputTree.GetBranch("GenPart_pdgId").GetTitle())
+        self.out.branch("GenPartCut_status",           "I", lenVar = "nGenPartCut", title=inputTree.GetBranch("GenPart_status").GetTitle())
+        self.out.branch("GenPartCut_statusFlags",      "I", lenVar = "nGenPartCut", title=inputTree.GetBranch("GenPart_statusFlags").GetTitle())
+
+    def endFile(self, inputFile, outputFile, inputTree, wrappedOutputTree):
+        pass
+
+    def recursiveMotherSearch(self, startIdx, GenPartCut_genPartIdxMother, filterArray):
+        if startIdx < 0: 
+            return startIdx
+
+        mom = GenPartCut_genPartIdxMother[startIdx]
+
+        if mom < 0 or filterArray[mom]:
+            return mom
+        else:
+            return self.recursiveMotherSearch(mom, GenPartCut_genPartIdxMother, filterArray)
+        
+
+    class TTreeReaderArrayWrapper:
+        def __init__(self, ttarray):
+            self.ttarray = ttarray
+
+        def __iter__(self):
+            for i in xrange(len(self.ttarray)):
+                yield self.ttarray[i]
+            return
+
+    def analyze(self, event):
+        """process event, return True (go to next module) or False (fail, go to next event)"""
+        ## Getting objects
+
+        GenPartCut_eta              = np.fromiter(self.TTreeReaderArrayWrapper(event.GenPart_eta),              dtype=float)
```

Comment:
Will you have the first event in file issue with this?

Reply:
no, because the np array creation is actually copying the data, but it seems to about cancel out with the speed boost using numpy.